File: 451_SortCharactersbyFrequency/451_SortCharactersbyFrequency.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function (s) {
  let cache = {};
  for (const letters of s) {
    cache[letters] ? cache[letters]++ : (cache[letters] = 1);
  }
  const vals = Object.keys(cache).sort((a, b) => cache[b] - cache[a]);
  let ans = '';
  for (let i = 0; i < vals.length; i++) {
    ans += vals[i].repeat(cache[vals[i]]);
  }
  return ans;
};

File: 451_SortCharactersbyFrequency/451_SortCharactersbyFrequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        cache = {}
        for letter in s:
            # python's version of a tertiary operator
            cache[letter] = cache[letter] + 1 if letter in cache else 1
            # array being sorted, anonymous function being applied, sorts the number in descending order
        vals = sorted(cache.keys(), key=lambda x: cache[x], reverse=True)
        ans = ''.join(char * cache[char] for char in vals)
    
        return ans